turbopack/crates/turbo-tasks-backend/src/data.rs

```diff
@@ -83,6 +85,140 @@ transient_traits!(RootState);
 
 impl Eq for RootState {}
 
+#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct DirtyState {
+    pub clean_in_session: Option<SessionId>,
+}
+
+impl DirtyState {
+    pub fn get(&self, session: SessionId) -> bool {
+        self.clean_in_session != Some(session)
+    }
+}
+
+fn add_with_diff(v: &mut i32, u: i32) -> i32 {
+    let old = *v;
+    *v += u;
+    if old <= 0 && *v > 0 {
+        1
+    } else if old > 0 && *v <= 0 {
+        -1
+    } else {
+        0
+    }
+}
+
+#[derive(Debug, Default, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
+pub struct DirtyContainerCount {
+    pub count: i32,
+    pub count_in_session: Option<(SessionId, i32)>,
+}
+
+impl DirtyContainerCount {
+    pub fn get(&self, session: SessionId) -> i32 {
+        if let Some((s, count)) = self.count_in_session {
+            if s == session {
+                return count;
+            }
+        }
+        self.count
+    }
+
+    pub fn update(&mut self, count: i32) -> DirtyContainerCount {
+        self.update_count(&DirtyContainerCount {
+            count,
+            count_in_session: None,
+        })
+    }
```

Comment:
I think it's risky for a `Copy` type to provide `&mut` methods. I could easily write some code that implicitly copies a value before mutating it, and not end up mutating what I thought I was mutating. People expect `Copy` types to be value types.

Either remove `Copy`, or treat this as a value type (make all these methods take ownership with `mut self` and return a new `DirtyContainerCount`).